src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+            if (expectedStoreTip.HashBlock != this.storeTip.HashBlock)
+                await this.RemoveReorgedBlocksFromStoreAsync(expectedStoreTip).ConfigureAwait(false);
+
+            // Save the batch.
+            ChainedHeader newTip = clearedBatch.Last().ChainedHeader;
+
+            this.logger.LogDebug("Saving batch of {0} blocks, total size: {1} bytes.", clearedBatch.Count, this.currentBatchSizeBytes);
+
+            await this.blockRepository.PutAsync(newTip.HashBlock, clearedBatch.Select(b => b.Block).ToList()).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+            this.logger.LogDebug("Store tip set to '{0}'.", this.storeTip);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Cleans the batch in a way that all headers from the latest one are consecutive.
+        /// Those that violate consecutiveness are removed.
+        /// </summary>
+        /// <param name="batch">Uncleaned batch that might contain non-consecutive blocks. Cannot be empty.</param>
+        /// <returns>List of consecutive blocks.</returns>
+        private List<BlockPair> GetBatchWithoutReorgedBlocks(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            // Initialize current with highest block from the batch.
+            BlockPair current = batch.Last();
+
+            // List of consecutive blocks. It's a cleaned out version of batch that doesn't have blocks that were reorged.
+            var batchCleared = new List<BlockPair>(batch.Count) { current };
+            
+            // Select only those blocks that were not reorged away.
+            for (int i = batch.Count - 2; i >= 0; i--)
+            {
+                if (batch[i].ChainedHeader.HashBlock != current.ChainedHeader.Previous.HashBlock)
+                {
+                    this.logger.LogDebug("Block '{0}' removed from the batch because it was reorged.", batch[i].ChainedHeader);
+                    continue;
+                }
+
+                batchCleared.Add(batch[i]);
+                current = batch[i];
+            }
+
+            batchCleared.Reverse();
+
+            this.logger.LogTrace("(-):*.{0}={1}", nameof(batchCleared.Count), batchCleared.Count);
+            return batchCleared;
+        }
+
+        /// <summary>Removes reorged blocks from the database.</summary>
+        /// <param name="expectedStoreTip">Highest block that should be in the store.</param>
+        private async Task RemoveReorgedBlocksFromStoreAsync(ChainedHeader expectedStoreTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(expectedStoreTip), expectedStoreTip);
+
+            var blocksToDelete = new List<uint256>();
+            ChainedHeader currentHeader = this.storeTip;
+
+            while (currentHeader.HashBlock != expectedStoreTip.HashBlock)
```

Comment:
Might we hit the bottom if `expectedStoreTip` is not in store.

Reply:
No, since expectedStoreTip is a prev block which should be in a store already. 

So unless block supplier sends blocks with gaps (so block's dont connect and we don't have a chain anymore) it will never happen. 

And if there are gaps- something is wrong on the block supplier side.